src/components/LanguagePicker/useLanguagePicker.tsx

```diff
@@ -0,0 +1,187 @@
+import { useEffect, useRef, useState } from "react"
+import { useRouter } from "next/router"
+import { useTranslation } from "next-i18next"
+import { useDisclosure, type UseDisclosureReturn } from "@chakra-ui/react"
+
+import type {
+  I18nLocale,
+  Lang,
+  LocaleDisplayInfo,
+  ProjectProgressData,
+} from "@/lib/types"
+
+import { MatomoEventOptions, trackCustomEvent } from "@/lib/utils/matomo"
+import { languages } from "@/lib/utils/translations"
+
+import progressData from "@/data/translationProgress.json"
+
+import { DEFAULT_LOCALE } from "@/lib/constants"
+
+const data = progressData as ProjectProgressData[]
+
+export const useLanguagePicker = (
+  handleClose?: () => void,
+  menuState?: UseDisclosureReturn
+) => {
+  const { t } = useTranslation("page-languages")
+  const { locale, locales } = useRouter()
+  const refs = {
+    inputRef: useRef<HTMLInputElement>(null),
+    firstItemRef: useRef<HTMLAnchorElement>(null),
+    noResultsRef: useRef<HTMLAnchorElement>(null),
+    footerRef: useRef<HTMLAnchorElement>(null),
+  }
+  const [filterValue, setFilterValue] = useState("")
+
+  const [filteredNames, setFilteredNames] = useState<LocaleDisplayInfo[]>([])
+
+  // perform all the filtering and mapping when the filter value change
+  useEffect(() => {
+    // Get the preferred languages for the users browser
+    const navLangs = typeof navigator !== "undefined" ? navigator.languages : []
+
+    // For each browser preference, reduce to the most specific match found in `locales` array
+    const allBrowserLocales: Lang[] = navLangs
+      .map(
+        (navLang) =>
+          locales?.reduce((acc, cur) => {
+            if (cur.toLowerCase() === navLang.toLowerCase()) return cur
+            if (
+              navLang.toLowerCase().startsWith(cur.toLowerCase()) &&
+              acc !== navLang
+            )
+              return cur
+            return acc
+          }, "") as Lang
+      )
+      .filter((i) => !!i) // Remove those without matches
+
+    // Remove duplicate matches
+    const browserLocales = Array.from(new Set(allBrowserLocales))
```

Comment:
Refactor idea: move all this logic to a util function.
```suggestion
    const browserLocales = getBrowserLocales(locales)
```